### PR TITLE
Storybook for dev server

### PR DESCRIPTION
**Pull request by dangodev**
## Reason for change
We need a more robust, future-proof docs system. If we do that, then the dev server has to change, because it currently is the docs.

Storybook lets us preview all states of our components at once. So we can focus on developing them, rather than focusing on presenting them in a docs format.

Screen Shot 2019-05-07 at 07 52 59

## Testing
Unfortunately Storybook won’t show up in the preview; you’ll have to run it locally. But this PR shouldn’t break the preview link (not yet).

### CODE REVIEW

File: .storybook/config.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  'manifold-plan-selector',
+  'manifold-toast',
+  'manifold-resource-credentials',
+  'manifold-resource-status',
+  'manifold-data-product-name',
+  'manifold-data-product-logo',
+  'manifold-data-resource-list',
+  'manifold-data-provision-button',
+  'manifold-data-manage-button',
+];
```
Comment (dangodev) [inline .storybook/config.js]: This is where we order the stories in the left-hand menu. I’m fine with doing alphabetical, but wanted to see what ordering would take.

File: .storybook/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    new CopyPlugin([
+      {
+        from: resolve(__dirname, '..', 'dist'),
+        to: resolve(__dirname, '..', 'node_modules', '@storybook', 'core', 'dist', 'public'),
+      },
+    ])
+  );
+
+  return config;
+};
```
Comment (dangodev) [inline .storybook/webpack.config.js]: This whole thing is a bit of a hack, but it gives us fast HMR in Storybook 🎉

File: readme.md [hunk 3]
```diff
@@ -1,5 +1,5 @@
+[![Built With Stencil](https://img.shields.io/badge/-Built%20With%20Stencil-16161d.svg?logo=data%3Aimage%2Fsvg%2Bxml%3Bbase64%2CPD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz4KPCEtLSBHZW5lcmF0b3I6IEFkb2JlIElsbHVzdHJhdG9yIDE5LjIuMSwgU1ZHIEV4cG9ydCBQbHVnLUluIC4gU1ZHIFZlcnNpb246IDYuMDAgQnVpbGQgMCkgIC0tPgo8c3ZnIHZlcnNpb249IjEuMSIgaWQ9IkxheWVyXzEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4IgoJIHZpZXdCb3g9IjAgMCA1MTIgNTEyIiBzdHlsZT0iZW5hYmxlLWJhY2tncm91bmQ6bmV3IDAgMCA1MTIgNTEyOyIgeG1sOnNwYWNlPSJwcmVzZXJ2ZSI%2BCjxzdHlsZSB0eXBlPSJ0ZXh0L2NzcyI%2BCgkuc3Qwe2ZpbGw6I0ZGRkZGRjt9Cjwvc3R5bGU%2BCjxwYXRoIGNsYXNzPSJzdDAiIGQ9Ik00MjQuNywzNzMuOWMwLDM3LjYtNTUuMSw2OC42LTkyLjcsNjguNkgxODAuNGMtMzcuOSwwLTkyLjctMzAuNy05Mi43LTY4LjZ2LTMuNmgzMzYuOVYzNzMuOXoiLz4KPHBhdGggY2xhc3M9InN0MCIgZD0iTTQyNC43LDI5Mi4xSDE4MC40Yy0zNy42LDAtOTIuNy0zMS05Mi43LTY4LjZ2LTMuNkgzMzJjMzcuNiwwLDkyLjcsMzEsOTIuNyw2OC42VjI5Mi4xeiIvPgo8cGF0aCBjbGFzcz0ic3QwIiBkPSJNNDI0LjcsMTQxLjdIODcuN3YtMy42YzAtMzcuNiw1NC44LTY4LjYsOTIuNy02OC42SDMzMmMzNy45LDAsOTIuNywzMC43LDkyLjcsNjguNlYxNDEuN3oiLz4KPC9zdmc%2BCg%3D%3D&colorA=16161d&style=flat-square)](https://stenciljs.com)
```
Comment (dangodev) [inline readme.md]: Found this badge on another site and I liked it better. Said the same thing 🤷‍♂

File: src/components/manifold-product-page/manifold-product-page.tsx [hunk 4]
```diff
@@ -65,7 +65,7 @@ export class ManifoldProductPage {
                 {this.hideCta !== true && (
                   <div class="sidebar-cta">
                     <manifold-link-button href={this.ctaLink} onClick={this.onClick}>
-                      <span class="link-text">Get {name} Duis mollis, est non commodo luctus</span>
+                      <span class="link-text">Get {name}</span>
```
Comment (dangodev) [inline src/components/manifold-product-page/manifold-product-page.tsx]: Remove the Dark Tongue of Mordor

File: stories/manifold-marketplace.stories.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+storiesOf('Marketplace', module)
+  .addParameters({ readme: { sidebar: markdown } })
+  .add('default', () => '<manifold-marketplace></manifold-marketplace>')
+  .add('no templates', () => '<manifold-marketplace hide-templates></manifold-marketplace>')
+  .add('compact', () => '<manifold-marketplace hide-categories></manifold-marketplace>')
+  .add(
+    'filtered',
+    () =>
+      '<manifold-marketplace hide-categories products="iron_cache,iron_mq,iron_worker"></manifold-marketplace>'
+  );
```
Comment (dangodev) [inline stories/manifold-marketplace.stories.js]: This is the basic story format.

`addParameters` loads the readme.

From there, every story follows the format `.add(storyName, HTML)`. It’s a super simple way to mock up many different components, and let us view them all quickly 🎉

**Comment by dangodev** [comment]
One thing I should note: @davidleger95 with Storybook there are no styles bleeding through, so on my machine many of the components are in Times New Roman. Rather than add global styles I think it’s probably best for you to tackle this in the theming PR, maybe?

File: src/components/manifold-product-page/manifold-product-page.tsx (hunk 4 above)
Comment (davidleger95) [reply src/components/manifold-product-page/manifold-product-page.tsx]: oops, already made a PR for this and merged it

**Comment by davidleger95** [comment]
I'm seeing that global stuff like the fonts aren't coming through. This might be fixed with proper theming but just wanted to call that out incase it should be there.

**Comment by davidleger95** [comment]
Overall, I think this approach is good. I'll have to go back to the thread in Slack and review the pros and cons you listed. Right now I'd maybe prefer docz but it's mostly because I'm curious about the stencil plugin. Maybe that's not a big advantage though.

**Comment by dangodev** [comment]
> I'll have to go back to the thread in Slack and review the pros and cons you listed. Right now I'd maybe prefer docz but it's mostly because I'm curious about the stencil plugin. Maybe that's not a big advantage though.

This wasn’t part of that; this would be the dev server, no matter what. That was the followup PR where we re-do the docs site. I think long-term separating docs from dev would be good. So I don’t really have an alternative to Storybook for dev. No matter what, we’ll have to do _something_ on Stencil dev, and IMO this is even less work than adding things to `index.html` by hand again.

**Comment by dangodev** [comment]
> I'm seeing that global stuff like the fonts aren't coming through. This might be fixed with proper theming but just wanted to call that out incase it should be there.

See my comment above (https://github.com/manifoldco/ui/pull/115#issuecomment-490101155). That was intentional, to expose what’s styled and what’s not.

When the new styling PR is good, then maybe we can pair on what the best way to expose that to Storybook would be.

File: stencil.config.ts [hunk 6]
```diff
@@ -8,7 +8,7 @@ export const config: Config = {
   namespace: 'manifold',
   globalStyle: 'src/global/theme.css',
   globalScript: 'src/global/app.ts',
-  outputTargets: [{ type: 'dist' }, { type: 'www', serviceWorker: null }],
```
Comment (dangodev) [inline stencil.config.ts]: Whoops just realized I can’t remove this yet until the new docs are up. Fixing 🔨

**Pull request merged**